### DISCUSSION TITLE
How to workaround the SVD issue with XLA

### DISCUSSION THREAD

**Discussion by pustovale** [NVlabs/sionna#112] [Q&A] [answered]
Hi all,

I implemented SVD-based precoding in Sionna, but encountered the issue which I'm struggling to resolve. 
The code adds SVD-precoding to the code of  "MIMO OFDM Transmissions over the CDL Channel Model" using the Tensorflow function:
`s, u, v = tf.linalg.svd(h)`

However, when running the code I encountered the error:
`No registered 'Svd' OpKernel for XLA_CPU_JIT`

I found that when disable XLA optimization for all the code 
```
@tf.function(jit_compile=False)
def call(self, batch_size, ebno_db):
```
everything works well. 
However, this way the code performance is degraded. I tried to workaround the SVD function with the same decorator  `@tf.function(jit_compile=False)` but it does not help to resolve the error if the rest of the code is run with enabled optimization.

Thus I have a question is there any other way to disable optimization for SVD while keeping the rest of the code with XLA.

**Comment by jhoydis** [comment, marked as answer]
Hi!

You can in principle put decorators around any part of the model you want. But you cannot put an XLA decorator around the entire model and then selectively exclude parts of the graph from it. So you would need to decorate specific components, e.g., the channel model and the receiver.

You could also try to implement SVD in a different manner. If an eigen decomposition would be sufficient, you could use [tf.linalg.eigh](https://www.tensorflow.org/api_docs/python/tf/linalg/eigh) which works with XLA. I am guessing that you might only need either ``u`` or ``v``. You could then probably do something like this:
```
_, u = tf.linalg.eigh(tf.matmul(h,h,adjoint_b=True))
_, v = tf.linalg.eigh(tf.matmul(h,h,adjoint_a=True))
```

**Reply by pustovale** [reply]
Thanks! The suggestion with `tf.linalg.eigh` works well.  For whom it may concern, don't forget to reverse the order of the columns in `v` with `v = tf.reverse(v,axis=[-1])`  as eigen values in `eigh` are sorted in the descend order, opposite to the `svd`.

**Comment by GCX-art** [comment]
Dear pustovale,
I hope this message finds you well.  I am currently working on a project involving Singular Value Decomposition (SVD) and its applications. I recently came across your work  and I found it both insightful and inspiring.

I am writing to seek your guidance regarding a specific issue I encountered while implementing SVD. I noticed that my implementation of SVD only performs well under low signal-to-noise ratio (SNR) conditions, and I am curious to understand why this might be the case. I was wondering if you could provide some insights into the potential reasons behind this behavior.Additionally, I would greatly appreciate it if you could share how you implemented SVD in your work. 

Thank you very much for your time and consideration. I understand that you must be very busy, so I truly appreciate any guidance or advice you can offer.

Looking forward to your response.

Best regards
![image](https://github.com/user-attachments/assets/b79308b8-9903-4c9d-8da3-4e0978d7c973)
